src/test/unit/mcmc/chainset_test.cpp

```diff
@@ -0,0 +1,334 @@
+    auto q_5 = chain_2.quantile(i + 7, 0.05);
+    EXPECT_NEAR(q_5, s8_q5(i), 0.5);
+    auto q_95 = chain_2.quantile(i + 7, 0.95);
+    EXPECT_NEAR(q_95, s8_q95(i), 0.5);
+    auto qs_5_50_95 = chain_2.quantiles(i + 7, probs);
+    EXPECT_EQ(3, qs_5_50_95.size());
+    EXPECT_NEAR(qs_5_50_95(0), s8_q5(i), 0.5);
+    EXPECT_NEAR(qs_5_50_95(1), s8_median(i), 0.05);
+    EXPECT_NEAR(qs_5_50_95(2), s8_q95(i), 0.5);
+  }
+}
+
+TEST_F(McmcChains, mcse) {
+  std::vector<stan::io::stan_csv> eight_schools;
+  eight_schools.push_back(eight_schools_1);
+  eight_schools.push_back(eight_schools_2);
+  stan::mcmc::chainset chain_2(eight_schools);
+  EXPECT_EQ(2, chain_2.num_chains());
+
+  // test against R implementation in pkg posterior
+  Eigen::VectorXd s8_mcse_mean(10), s8_mcse_sd(10);
+  s8_mcse_mean << 0.288379, 0.4741815, 0.2741001, 0.3294614, 0.2473758,
+      0.2665048, 0.2701363, 0.4740092, 0.3621771, 0.3832464;
+  s8_mcse_sd << 0.1841825, 0.2854258, 0.192332, 0.2919369, 0.2478025, 0.2207478,
+      0.2308452, 0.2522107, 0.2946896, 0.3184745;
+
+  for (size_t i = 0; i < 10; ++i) {
+    auto mcse_mean = chain_2.mcse_mean(i + 7);
+    EXPECT_NEAR(mcse_mean, s8_mcse_mean(i), 0.5);
+    auto mcse_sd = chain_2.mcse_sd(i + 7);
+    EXPECT_NEAR(mcse_sd, s8_mcse_sd(i), 0.7);
+  }
+}
+
+TEST_F(McmcChains, const_fail) {
+  std::ifstream bernoulli_const_1_stream, bernoulli_const_2_stream;
+  stan::io::stan_csv bernoulli_const_1, bernoulli_const_2;
+  bernoulli_const_1_stream.open(
+      "src/test/unit/mcmc/test_csv_files/bernoulli_const_1.csv",
+      std::ifstream::in);
+  bernoulli_const_1
+      = stan::io::stan_csv_reader::parse(bernoulli_const_1_stream, &out);
+  bernoulli_const_1_stream.close();
+  bernoulli_const_2_stream.open(
+      "src/test/unit/mcmc/test_csv_files/bernoulli_const_2.csv",
+      std::ifstream::in);
+  bernoulli_const_2
+      = stan::io::stan_csv_reader::parse(bernoulli_const_2_stream, &out);
+  bernoulli_const_2_stream.close();
+  std::vector<stan::io::stan_csv> bernoulli_const;
+  bernoulli_const.push_back(bernoulli_const_1);
+  bernoulli_const.push_back(bernoulli_const_2);
+  stan::mcmc::chainset chain_2(bernoulli_const);
+  EXPECT_EQ(2, chain_2.num_chains());
+  auto rhat = chain_2.split_rank_normalized_rhat("zeta");
+  EXPECT_TRUE(std::isnan(rhat.first));
+  EXPECT_TRUE(std::isnan(rhat.second));
+  auto ess = chain_2.split_rank_normalized_ess("zeta");
+  EXPECT_TRUE(std::isnan(ess.first));
+  EXPECT_TRUE(std::isnan(ess.second));
+}
+
+TEST_F(McmcChains, autocorrelation) {
+  stan::mcmc::chainset chain_1(eight_schools_1);
+  EXPECT_EQ(1, chain_1.num_chains());
+
+  Eigen::VectorXd mu_ac_posterior(10);
+  mu_ac_posterior << 1.00000000000, 0.19487668999, 0.05412049365, 0.07834048575,
+      0.04145609855, 0.04353962161, -0.00977255885, 0.00005175308,
+      0.01791577080, 0.01245035817;
+  auto mu_ac = chain_1.autocorrelation(0, "mu");
+  for (size_t i = 0; i < 10; ++i) {
+    EXPECT_NEAR(mu_ac_posterior(i), mu_ac(i), 0.05);
```

Comment:
Tolerance of 0.05 when expected answer is 0.00005 -- no good!

Reply:
made test more strict, but differences between R and C++ are to be expected.